codec-http/src/main/java/io/netty/handler/codec/http/router/Router.java

```diff
@@ -0,0 +1,551 @@
+     * {@link #route(HttpMethod, String)}.
+     */
+    public Router notFound(T target) {
+        this.notFound = target;
+        return this;
+    }
+
+    private MethodlessRouter<T> getMethodlessRouter(HttpMethod method) {
+        if (method == null) {
+            return anyMethodRouter;
+        }
+
+        MethodlessRouter<T> r = routers.get(method);
+        if (r == null) {
+            r = new MethodlessRouter<T>();
+            routers.put(method, r);
+        }
+
+        return r;
+    }
+
+    //--------------------------------------------------------------------------
+
+    /** Removes the route specified by the path. */
+    public void removePath(String path) {
+        for (MethodlessRouter<T> r : routers.values()) {
+            r.removePath(path);
+        }
+        anyMethodRouter.removePath(path);
+    }
+
+    /** Removes all routes leading to the target. */
+    public void removeTarget(T target) {
+        for (MethodlessRouter<T> r : routers.values()) {
+            r.removeTarget(target);
+        }
+        anyMethodRouter.removeTarget(target);
+    }
+
+    //--------------------------------------------------------------------------
+
+    /**
+     * If there's no match, returns the result with {@link #notFound(Object) notFound}
+     * as the target if it is set, otherwise returns {@code null}.
+     */
+    public RouteResult<T> route(HttpMethod method, String uri) {
+        QueryStringDecoder decoder = new QueryStringDecoder(uri);
+        String[]           tokens  = Path.removeSlashesAtBothEnds(decoder.path()).split("/");
+
+        MethodlessRouter<T> router = routers.get(method);
+        if (router == null) {
+            router = anyMethodRouter;
+        }
+
+        RouteResult<T> ret = router.route(tokens);
+        if (ret != null) {
+            return new RouteResult(ret.target(), ret.pathParams(), decoder.parameters());
+        }
+
+        if (router != anyMethodRouter) {
+            ret = anyMethodRouter.route(tokens);
+            if (ret != null) {
+                return new RouteResult(ret.target(), ret.pathParams(), decoder.parameters());
+            }
+        }
+
+        if (notFound != null) {
+            // Return mutable map to be consistent, instead of
+            // Collections.<String, String>emptyMap()
+            return new RouteResult<T>(notFound, new HashMap<String, String>(), decoder.parameters());
+        }
+
+        return null;
+    }
+
+    //--------------------------------------------------------------------------
+    // For implementing OPTIONS and CORS.
+
+    /**
+     * Returns allowed methods for a specific URI.
+     *
+     * For {@code OPTIONS *}, use {@link #allAllowedMethods()} instead of this method.
+     */
+    public Set<HttpMethod> allowedMethods(String uri) {
+        QueryStringDecoder decoder = new QueryStringDecoder(uri);
+        String[]           tokens  = Path.removeSlashesAtBothEnds(decoder.path()).split("/");
```

Comment:
StringUtil looks great. Any plan to move it from `io.netty.util.internal` to `io.netty.util`?